shared/lib_battery.cpp

```diff
@@ -419,17 +450,24 @@ double capacity_kibam_t::q20(){return _q20;}
 Define Lithium Ion capacity model
 */
 capacity_lithium_ion_t::capacity_lithium_ion_t() { /* nothing to do */ }
-capacity_lithium_ion_t::capacity_lithium_ion_t(double q, double SOC_init, double SOC_max, double SOC_min) :capacity_t(q, SOC_init, SOC_max, SOC_min){};
+capacity_lithium_ion_t::capacity_lithium_ion_t(double q, double SOC_init, double SOC_max, double SOC_min, double dt_hr):
+capacity_t(q, SOC_init, SOC_max, SOC_min){
+    _dt_hour = dt_hr;
+
+};
 capacity_lithium_ion_t * capacity_lithium_ion_t::clone(){ return new capacity_lithium_ion_t(*this); }
 void capacity_lithium_ion_t::copy(capacity_t * capacity){ capacity_t::copy(capacity);}
 
 void capacity_lithium_ion_t::replace_battery(double replacement_percent)
 {
+    replacement_percent = fmax(0, replacement_percent);
+    double qmax_old = _qmax;
 	_qmax += _qmax0 * replacement_percent * 0.01;
 	_qmax = fmin(_qmax0, _qmax);
-	_q0 = _qmax * _SOC_init * 0.01;
 	_qmax_thermal = _qmax;
+	_q0 += (_qmax-qmax_old) * _SOC_init * 0.01;
```

Comment:
confused by this- why are you subtracting the previous qmax here? I see it in a couple places.

Reply:
So in this, I assumed the idea with the replacement_percent x was that we're replacing up to x percent of the battery. So if the current max capacity has degraded down to 50% of original and the replacement percent is 70, then we'll be adding a new 20% of original capacity. Hence, the capacity would increase by (70-50) * initial SOC.

I couldn't find much info on exactly what is meant in replacing the battery. But another interpretation would be to replace that 50% plus an additional 20%. Then the capacity would be equal to 70 * initial SOC.

Now that you've mentioned it, I think the second interpretation makes more sense. Especially with how replacement costs are probably calculated using that 70% rather than the 20% new capacity. What do you think?

Reply:
I think I'll change it so that if you specify replacement of 70% that means you get rid of the old battery entirely and add a new battery of 70% of original battery

Reply:
Well actually the original code does this, where qmax0 is the nominal max capacity

```
_qmax += replacement_percent * 0.01* _qmax0;
_qmax = fmin(_qmax, _qmax0);
_q0 = _qmax*_SOC_init*0.01;
```
So we're adding to the degraded capacity some percent of the nominal, up to 100% of the nominal. This replacement follows a schedule or by a minimum capacity after which replacement occurs. 

However the replacement costs are not actually calculated using how much replacement has occurred. In lib_battery_dispatch.cpp 1613:

```
double capacityPercentDamagePerCycle = _Battery->lifetime_model()->cycleModel()->estimateCycleDamage();
m_cycleCost = 0.01 * capacityPercentDamagePerCycle * m_battReplacementCostPerKWH;
```
where estimateCycleDamage estimates the capacity lost in running the next cycle, whose range is the same as the average range thus far. This is a very different quantity than how much ACTUAL replacement has occurred * the cost to replace.

What I'm proposing:
1. track how much capacity is actually replaced
2. cycling cost = actual replaced capacity * cost per kwh

Reply:
Hmm. Maybe we can ask around about what it means to replace part of a battery. I don't know which of your proposed capacity calculations (or the original definition) make the most sense. But I do think that your proposed cycle cost makes sense.